src/ucp/core/ucp_worker.c

```diff
@@ -2345,32 +2363,80 @@ static unsigned ucp_worker_discard_uct_ep_progress(void *arg)
-                                            void *arg)
+static int
+ucp_worker_discard_remove_filter(const ucs_callbackq_elem_t *elem, void *arg)
 {
-    if ((elem->cb != ucp_worker_discard_uct_ep_destroy_progress) &&
-        (elem->cb != ucp_worker_discard_uct_ep_progress)) {
+    if ((elem->arg != arg) ||
+        ((elem->cb != ucp_worker_discard_uct_ep_destroy_progress) &&
+         (elem->cb != ucp_worker_discard_uct_ep_progress))) {
         return 0;
     }
 
-    ucp_worker_put_flush_req((ucp_request_t*)elem->arg);
+    ucp_worker_discard_uct_ep_complete((ucp_request_t*)elem->arg);
     return 1;
 }
 
-static void ucp_worker_discarded_uct_eps_cleanup(ucp_worker_h worker)
+static void
+ucp_worker_discard_uct_ep_pending_count_purged(uct_pending_req_t *self,
+                                               void *arg)
 {
+    ++(*((unsigned*)arg));
+}
+
+static void ucp_worker_discard_uct_ep_cleanup(ucp_worker_h worker)
+{
+    ucp_request_t *req;
+    unsigned num_purged;
     uct_ep_h uct_ep;
+    UCS_HLIST_HEAD(discard_reqs);
+
+    /* Destroying UCP EP in ucp_ep_disconnected() could start UCT EP discarding
+     * operations (e.g. for TMP EP in CM lane). Do cleanup of discarding
+     * functionality after trying to destroy UCP EPs in order to destroy all
+     * remaining UCP EPs here */
+
+    kh_foreach(&worker->discard_uct_ep_hash, uct_ep, req, {
+        /* Save all UCP requests from the discarding hash to a temporal list and
+         * then go through all list entries and cleanup discard requests, since
+         * khash doesn't have robust iterator which allows remove removing its
+         * elements while iterating */
+        ucs_assert(uct_ep == req->send.discard_uct_ep.uct_ep);
+        ucs_hlist_add_tail(&discard_reqs, &req->send.list);
+    })
 
-    ucs_callbackq_remove_if(&worker->uct->progress_q,
-                            ucp_worker_discard_remove_filter, NULL);
+    ucs_hlist_for_each_extract(req, &discard_reqs, send.list) {
+        num_purged = 0;
+        uct_ep     = req->send.discard_uct_ep.uct_ep;
 
-    /* If ep owns the discard operation ep_destroy will cancel it. We are after
-     * uct_worker_progress_unregister_safe and ucp_worker_discard_remove_filter,
-     * so either we canceled req or it was finished and removed from kh before
-     */
-    kh_foreach_key(&worker->discard_uct_ep_hash, uct_ep, {
+        uct_ep_pending_purge(uct_ep,
+                             ucp_worker_discard_uct_ep_pending_count_purged,
+                             &num_purged);
         uct_ep_destroy(uct_ep);
-    })
+
+        if (num_purged == 0) {
+            /* Either flush is in-progress (will be completed in UCT EP destroy)
+             * or some progress callback is scheduled (will be removed after UCT
+             * EP destroy) */
+            ucs_callbackq_remove_if(&worker->uct->progress_q,
+                                    ucp_worker_discard_remove_filter, req);
```

Comment:
to avoid extra list, we can avoid removing from hash during iteration by moving kh_del coude out of ucp_worker_discard_uct_ep_complete() (revert)

Reply:
fixed